ribs/archives/_nearest_neighbor_archive.py

```diff
@@ -0,0 +1,489 @@
+              the archive was empty), the novelty is set to infinity
+              (``numpy.inf``).
+
+        Raises:
+            ValueError: The array arguments do not match their specified shapes.
+            ValueError: ``objective`` or ``measures`` has non-finite values (inf
+                or NaN).
+        """
+        data = validate_batch(
+            self,
+            {
+                "solution":
+                    solution,
+                "objective":
+                    np.zeros(len(solution), dtype=self.dtypes["objective"])
+                    if objective is None else objective,
+                "measures":
+                    measures,
+                **fields,
+            },
+        )
+
+        reference_measures = self.data("measures")
+        if self._compare_to_batch:
+            reference_measures = np.concatenate(
+                (reference_measures, data["measures"]), axis=0)
+
+        if self.compare_to_batch:
+            # In this case, solutions from the batch will be included in the
+            # nearest neighbors. Thus, the first neighbor of each solution will
+            # be the solution itself, so we need to compute an extra neighbor
+            # and exclude the first neighbor.
+            k_neighbors = min(len(reference_measures), self.k_neighbors + 1)
+            first_neighbor_idx = 1
+            kd_tree = cKDTree(reference_measures, **self._ckdtree_kwargs)
+        else:
+            k_neighbors = min(len(reference_measures), self.k_neighbors)
+            first_neighbor_idx = 0
+            kd_tree = self._cur_kd_tree
+
+        if len(reference_measures) == 0:
+            # If there are no references for computing nearest neighbors, there
+            # is infinite novelty and all solutions are added.
+            novelty = np.full(len(data["measures"]),
+                              np.inf,
+                              dtype=self.dtypes["measures"])
+            eligible = np.ones(len(data["measures"]), dtype=bool)
+        else:
+            # Compute nearest neighbors.
+            dists, _ = kd_tree.query(data["measures"], k=k_neighbors)
+
+            # Expand since query() automatically squeezes the last dim when k=1.
+            dists = dists[:, None] if k_neighbors == 1 else dists
+
+            novelty = np.mean(dists[:, first_neighbor_idx:], axis=1)
+            eligible = novelty >= self.novelty_threshold
+
+        n_eligible = np.sum(eligible)
+        new_size = len(self) + n_eligible
+
+        if new_size > self.capacity:
+            # Resize the store by doubling its capacity. We may need to double
+            # the capacity multiple times. The log2 below indicates how many
+            # times we would need to double the capacity. We obtain the final
+            # multiplier by raising to a power of 2.
+            multiplier = 2**int(np.ceil(np.log2(new_size / self.capacity)))
+            self._store.resize(multiplier * self.capacity)
```

Comment:
it is possible that the user sets the initial capacity to 0, therefore the archive would never grow using this strategy.

I think it's simplest to just throw an `initial_capacity = max(1, initial_capacity)` in `__init__`

Reply:
Good point! I now throw an error if initial_capacity is less than 1.